### PR TITLE
Add build pricing and item optimization to Items Tab

### PR DESCRIPTION
**Pull request by Nostrademous**
Avert thy eyes.....

TODOs:
- [x] Fix hardcode league name to different stored location (probably best as part of the league selection drop down)
- [x] Make Weighted Sum searches always sort by descending weighted sum
- [x] Add tooltip for URL EditControl to state that CTRL+CLICK will open browser and that browser don't respect encoded sorting specification (GGG is aware, might fix in future) so web-base item list will differ in ordering

### CODE REVIEW

File: src/Classes/TradeQuery.lua [hunk 1]
```diff
@@ -0,0 +1,897 @@
...
+			self.pbFileTimestampDiff[controls.league.selIndex] = get_time() - self.lastCurrencyFileTime[controls.league.selIndex] 
+		end
+		if self.pbFileTimestampDiff[controls.league.selIndex] == nil or self.pbFileTimestampDiff[controls.league.selIndex] >= 3600 then
+			tooltip:AddLine(16, colorCodes.WARNING .. "Currency Conversion rates are pulled from PoE Ninja leveraging their API.")
+			tooltip:AddLine(16, colorCodes.WARNING .. "Updates are limited to once per hour and not necessary more than once per day.")
+			tooltip:AddLine(16, "")
+			tooltip:AddLine(16, colorCodes.NEGATIVE .. "NOTE: This will expose your IP address to poe.ninja.")
+			tooltip:AddLine(16, colorCodes.NEGATIVE .. "If you are concerned about this please do not click this button.")
+		elseif self.pbFileTimestampDiff[controls.league.selIndex] ~= nil and self.pbFileTimestampDiff[controls.league.selIndex] < 3600 then
+			tooltip:AddLine(16, "Conversion Rates are less than an hour old (" .. tostring(self.pbFileTimestampDiff[controls.league.selIndex]) .. " seconds old)")
```
Comment (Wires77) [inline src/Classes/TradeQuery.lua:437]: Ideally this would update to minutes at some point (doesn't need second-by-second granularity after the first minute), since 12345 seconds is hard to read.

File: src/Classes/TradeQuery.lua [hunk 2]
```diff
@@ -0,0 +1,897 @@
...
+	end
+	controls['importButton'..str_cnt].enabled = function()
+		return #controls['importButtonText'..str_cnt].buf > 0
+	end
+	-- for storing the base64 item description
+	controls['importButtonText'..str_cnt] = new("EditControl", nil, 0, 0, 0, 0, "", nil, "", nil, nil, 16)
+	controls['importButtonText'..str_cnt].shown = false
+	-- for storing the whisper to purchase item from seller
+	controls['whisperButtonText'..str_cnt] = new("EditControl", nil, 0, 0, 0, 0, "", nil, "", nil, nil, 16)
+	controls['whisperButtonText'..str_cnt].shown = false
```
Comment (Wires77) [inline src/Classes/TradeQuery.lua:862]: Why are we storing text in hidden UI controls?  Seems like a terrible code smell and there has to be a better data structure that can hold them.  The results end up getting wiped if you close the modal also, which means all of the slots have to be searched again.

File: src/Classes/TradeQuery.lua [hunk 3]
```diff
@@ -0,0 +1,897 @@
+-- Path of Building
```
Comment (Wires77) [inline src/Classes/TradeQuery.lua:1]: I haven't fully vetted the queries yet, but there is a lot that can be cleaned up about this file.  For starters, having separate classes for each API would be a huge boon to readability.  Mixing the UI elements with backend calls in one file is really hard to follow.

Secondly, there are better ways to display this data than in EditControls.  Simple text works fine and can be updated easily.  Maybe for the status indicator you can change the text of "Price Item" or have a tooltip pop-up over the disabled button, I'm not sure.  Having a random edit control at the top is confusing, though.

File: src/Classes/TradeQuery.lua [hunk 4]
```diff
@@ -0,0 +1,897 @@
...
+	end
+	controls.leagueLabel = new("LabelControl", {"TOPRIGHT",controls.league,"TOPLEFT"}, -4, 0, 20, 16, "League:")
+	controls.poesessidButton = new("ButtonControl", {"TOPLEFT",controls.leagueLabel,"TOPLEFT"}, -256, 0, 240, row_height, POESESSID ~= "" and "HAVE POESESSID" or colorCodes.WARNING.."NEED POESESSID", function()
+		local poesessid_controls = {}
+		poesessid_controls.sessionInput = new("EditControl", nil, 0, 18, 350, 18, #POESESSID == 32 and POESESSID or "<PASTE POESESSID FROM BROWSER>", "POESESSID", "%X", 32, function(buf)
+			if #poesessid_controls.sessionInput.buf == 32 then
+				POESESSID = poesessid_controls.sessionInput.buf
+				controls.poesessidButton.label = "HAVE POESESSID"
+			end
+		end)
```
Comment (Wires77) [inline src/Classes/TradeQuery.lua:364]: I was somehow able to break this by trying to import from an account that doesn't exist in the ImportTab, coming back to this modal, and it telling me I "HAVE POESESSID"

File: src/Classes/TradeQuery.lua [hunk 5]
```diff
@@ -0,0 +1,897 @@
...
+
+-- Method to pull down and interpret available leagues from PoE
+function TradeQueryClass:PullLeagueList(controls)
+	local id = LaunchSubScript([[
+		local curl = require("lcurl.safe")
+		local page = ""
+		local easy = curl.easy()
+		easy:setopt{
+			url = "https://api.pathofexile.com/leagues?type=main&compact=1",
+			httpheader = {'Content-Type: application/json', 'Accept: application/json', 'User-Agent: Path of Building/]]..launch.versionNumber..[[ (contact: pob@mailbox.org)'}
```
Comment (Wires77) [inline src/Classes/TradeQuery.lua:146]: Does `pob@mailbox.org` actually exist or is it just a placeholder?  Might want to change that to the GitHub URL if not.

File: src/Data/Global.lua [hunk 6]
```diff
@@ -284,3 +284,5 @@ GlobalCache = {
 	useFullDPS = false,
 	numActiveSkillInFullDPS = 0,
 }
+
+POESESSID = ""
```
Comment (Wires77) [inline src/Data/Global.lua:288]: This could be better suited to being saved in Settings.xml.  Obviously it can expire, but maybe we can prompt for an update then?

File: src/Classes/ImportTab.lua [hunk 7]
```diff
@@ -313,7 +313,7 @@ function ImportTabClass:DownloadCharacterList()
 	  -- Trim Trailing/Leading spaces
 	local accountName = self.controls.accountName.buf:gsub('%s+', '')
 	local realm = realmList[self.controls.accountRealm.selIndex]
-	local sessionID = #self.controls.sessionInput.buf == 32 and self.controls.sessionInput.buf or (main.gameAccounts[accountName] and main.gameAccounts[accountName].sessionID)
+	POESESSID = #self.controls.sessionInput.buf == 32 and self.controls.sessionInput.buf or (main.gameAccounts[accountName] and main.gameAccounts[accountName].sessionID)
```
Comment (Wires77) [inline src/Classes/ImportTab.lua:316]: I would love for this tab to share session input logic with the new trader modal if possible.  Having one UI show the instructions as text and the other as a tooltip in a modal doesn't feel great.

File: src/Classes/TradeQueryGenerator.lua [hunk 8]
```diff
@@ -0,0 +1,695 @@
...
+}
+
+local tradeStatCategoryIndices = {
+    ["Explicit"] = 2,
+    ["Implicit"] = 3,
+    ["Corrupted"] = 3,
+    ["Scourge"] = 6
+}
+
+local influenceSuffixes = { "_shaper", "_elder", "_adjudicator", "_basilisk", "_crusader", "_eyrie"}
```
Comment (Wires77) [inline src/Classes/TradeQueryGenerator.lua:35]: These already exist in ItemTools.lua

File: src/Classes/TradeQueryGenerator.lua [hunk 9]
```diff
@@ -0,0 +1,695 @@
...
+            elseif statOrder == 1725 then
+                specialCaseData.overrideModLineSingular = "You can apply an additional Curse"
+                if modLine == specialCaseData.overrideModLineSingular then
+                    modLine = "You can apply 1 additional Curses"
+                end
+            elseif statOrder == 1366 then
+                specialCaseData.overrideModLineSingular = "Bow Attacks fire an additional Arrow"
+                if modLine == specialCaseData.overrideModLineSingular then
+                    modLine = "Bow Attacks fire 1 additional Arrows"
+                end
```
Comment (Wires77) [inline src/Classes/TradeQueryGenerator.lua:148]: Are we sure these magic numbers are going to stay the same?

File: src/Classes/TradeQueryGenerator.lua [hunk 10]
```diff
@@ -0,0 +1,695 @@
...
+        elseif mod.weightVal[i] > 0 and tags[influenceStrippedKey] == true then
+            return true
+        end
+    end
+    return false
+end
+
+function TradeQueryGeneratorClass:GenerateModData(mods, tradeQueryStatsParsed)
+    for modId, mod in pairs(mods) do
+        if localOnlyModGroups[mod.group] == true or (modId:find("Local") ~= nil and modId:find("Socketed") == nil) then -- skip all local mods other than socket level mods
```
Comment (Wires77) [inline src/Classes/TradeQueryGenerator.lua:119]: You might be able to parse locality from the ggpk.  I just don't think we save it off into ModItem.lua

File: src/Classes/TradeQueryGenerator.lua [hunk 11]
```diff
@@ -0,0 +1,695 @@
...
+            if #tokens ~= 0 and #tokens ~= 2 and #tokens ~= 4 then
+                logToFile("Unexpected # of tokens found for mod: %s", mod[i])
+                goto nextModLine
+            end
+
+            -- Update the min and max values available for each item category
+            for category, tags in pairs(itemCategoryTags) do
+                if canModSpawnForItemCategory(mod, tags) then
+                    if self.modData[modType][statOrder][category] == nil then
+                        self.modData[modType][statOrder][category] = { min = 999999, max = -999999 }
```
Comment (Wires77) [inline src/Classes/TradeQueryGenerator.lua:202]: math.max() and math.min() might be better suited here?

File: src/Classes/TradeQueryGenerator.lua [hunk 12]
```diff
@@ -0,0 +1,695 @@
...
+                        modRange.max = math.max(modRange.max, tokens[2])
+                    elseif #tokens == 4 then
+                        modRange.min = math.min(modRange.min, (tokens[1] + tokens[3]) / 2)
+                        modRange.max = math.max(modRange.max, (tokens[2] + tokens[4]) / 2)
+                    end
+                end
+            end
+            ::nextModLine::
+        end
+        ::continue::
```
Comment (Wires77) [inline src/Classes/TradeQueryGenerator.lua:220]: Could break these loops into their own functions and just return from them vs. using jumps.  Would probably help readability, too.

File: src/Classes/TradeQueryGenerator.lua [hunk 13]
```diff
@@ -0,0 +1,695 @@
...
+
+function TradeQueryGeneratorClass:StartQuery(slot, options)
+    -- Figure out what type of item we're searching for
+    local existingItem = self.itemsTab.items[slot.selItemId]
+    local testItemType = existingItem and existingItem.baseName or "Unset Amulet"
+    local itemCategoryQueryStr
+    local itemCategory
+    if slot.slotName == "Weapon 2" or slot.slotName == "Weapon 1" then
+        if existingItem.type == "Shield" then
+            itemCategoryQueryStr = "armour.shield"
```
Comment (Wires77) [inline src/Classes/TradeQueryGenerator.lua:424]: This string of ifs might be more appropriate as a table.  Ideally generated, but I'm not sure if that is easy to do.

File: src/Classes/TradeQuery.lua [hunk 14]
```diff
@@ -0,0 +1,897 @@
...
+	controls['uri'..str_cnt].tooltipFunc = function(tooltip)
+		tooltip:Clear()
+		if controls['uri'..str_cnt].buf:find('^https://www.pathofexile.com/trade/search/') ~= nil then
+			tooltip:AddLine(16, "CTRL click to open in web-browser or click 'Price Item' to do it in PoB")
+			tooltip:AddLine(16, "")
+			tooltip:AddLine(14, colorCodes.NEGATIVE .. "NOTE: you will need to re-sort until GGG fixes")
+		end
+	end
+	controls['priceButton'..str_cnt] = new("ButtonControl", {"TOPLEFT",controls['uri'..str_cnt],"TOPLEFT"}, 500 + 8, 0, 100, row_height, "Price Item", function()
+		self:PublicTrade(controls['uri'..str_cnt].buf, slotTbl, controls, str_cnt)
```
Comment (Wires77) [inline src/Classes/TradeQuery.lua:804]: Definitely disable the button right away instead of waiting for the query first

File: src/Classes/TradeQuery.lua [hunk 15]
```diff
@@ -0,0 +1,897 @@
...
+	--	 Up to 16 fetch requests in a 14 second window
+	self.rate_short_fetch_window = { }
+	self.rate_short_fetch_time = 6
+	-- we reduce from 12 to 7 since we may want up to 5 fetch slots for each request
+	self.rate_short_max_fetches = 7
+	self.rate_long_fetch_window = { }
+	self.rate_long_fetch_time = 14
+	-- we reduce from 16 to 11 since we may want up to 5 fetch slots for each request
+	self.rate_long_max_fetches = 11
+
```
Comment (Wires77) [inline src/Classes/TradeQuery.lua:116]: These need to come from the GGG API response, not be hardcoded.  See https://www.pathofexile.com/developer/docs/index#ratelimits

Following those docs would also eliminate some of the helper methods as well.

**Review by Wires77 — CHANGES_REQUESTED** [review]
Submitting this review for now.  Haven't fully reviewed the TradeQueryGenerator (I couldn't get it to do anything, actually) or the API calls yet, but I made some UI suggestions at least.

I think users will be overwhelmed with information as it currently looks.  Hiding things that aren't needed until after a query would be a good start, as well as trimming down the explanatory text.  Styling is all over the place, but that's an easy fix later on.  Oh, and of course the modal doesn't work in portrait mode 🥲 

It's shaping up, but we still have a lot to iterate on before it's ready for beta.

Actual error screens I ran into while testing:

**Review by Wires77 — COMMENTED** [review]
Reviewed the TradeQueryGenerator more in-depth.  The scourge bits could be contingent on the selected league, the API calls could be pulled out to a general PoE API class (see other comments), and some of the functions could benefit from helper functions.  Overall this looks pretty good.  

I'd love to see this changed to be more testable and have some unit tests written for at least some of the helper functions that calculate weight and such, as it seems like a great candidate for that.

File: src/Classes/TradeQuery.lua [hunk 16]
```diff
@@ -0,0 +1,897 @@
...
+    -- set 
+    self.storedGlobalCacheDPSView = GlobalCache.useFullDPS
+	GlobalCache.useFullDPS = GlobalCache.numActiveSkillInFullDPS > 0
+end)
+
+-- Method to process JSON to Lua translation
+function TradeQueryClass:ProcessJSON(json)
+	return dkjson.decode(json)
+end
```
[the remark this answers is not in the document]
Comment (Nostrademous) [reply src/Classes/TradeQuery.lua:136]: It's useful to keep like this when you need to debug a JSON change. Only reason.

File: src/Classes/TradeQuery.lua [hunk 17]
```diff
@@ -0,0 +1,897 @@
...
+				self:SetCurrencyConversionButton(controls)
+			end
+		end)
+	end
+end
+
+-- Method to set last PoE.Ninja endpoint access
+function TradeQueryClass:PriceBuilderPoENinjaCurrencyRequest()
+	self.lastCurrencyConversionRequest = get_time()
+end
```
[the remark this answers is not in the document]
Comment (Nostrademous) [reply src/Classes/TradeQuery.lua:194]: I left it like that incase PoE Ninja had issues with frequency of PoB users hitting his site for this and requested that we add some more elaborate timeout. You can easily add an if-else or other offset here.

File: src/Classes/TradeQuery.lua [hunk 18]
```diff
@@ -0,0 +1,897 @@
...
+	controls['resultCount'..str_cnt] = new("EditControl", {"TOPLEFT",controls['resultIndex'..str_cnt],"TOPLEFT"}, 60 + 8, 0, 82, row_height, "^8No Results", nil, nil)
+	controls['priceAmount'..str_cnt] = new("EditControl", {"TOPLEFT",controls['resultCount'..str_cnt],"TOPLEFT"}, 82 + 8, 0, 120, row_height, "Price", nil, nil)
+	controls['priceAmount'..str_cnt].enabled = function()
+		local boolean = #controls['priceAmount'..str_cnt].buf > 0
+		return boolean
+	end
+	controls['importButton'..str_cnt] = new("ButtonControl", {"TOPLEFT",controls['priceAmount'..str_cnt],"TOPLEFT"}, 120 + 8, 0, 100, row_height, "Import Item", function()
+		self.itemsTab:CreateDisplayItemFromRaw(controls['importButtonText'..str_cnt].buf)
+		local item = self.itemsTab.displayItem
+		-- pass "true" to not auto equip it as we will have our own logic
```
[the remark this answers is not in the document]
Comment (Nostrademous) [reply src/Classes/TradeQuery.lua:836]: Because there are three main differences in how it's implemented:
1) I don't want to create an extra click in to "save" the item
2) I don't want to leave the item in the preview-pane requiring an "Add To Build" click
3) I want the item to auto-populate in the right slot when the button is hit (even when an Item is already in that slot)
3a) I differentiate between Left and Right Ring Slots (for example) as well. The default doesn't.

File: src/Classes/TradeQuery.lua (hunk 1 above)
Comment (Nostrademous) [reply src/Classes/TradeQuery.lua:437]: Sure, that could be a generic function when dealing with time to do proper conversion.

File: src/Classes/TradeQuery.lua [hunk 19]
```diff
@@ -0,0 +1,897 @@
...
+currencyConversionTradeMap["Orb of Alchemy"] = "alch"
+currencyConversionTradeMap["Chaos Orb"] = "chaos"
+currencyConversionTradeMap["Gemcutter's Prism"] = "gcp"
+currencyConversionTradeMap["Exalted Orb"] = "exalted"
+currencyConversionTradeMap["Chromatic Orb"] = "chrome"
+currencyConversionTradeMap["Jeweller's Orb"] = "jewellers"
+currencyConversionTradeMap["Engineer's Orb"] = "engineers"
+currencyConversionTradeMap["Infused Engineer's Orb"] = "infused-engineers-orb"
+currencyConversionTradeMap["Orb of Chance"] = "chance"
+currencyConversionTradeMap["Cartographer's Chisel"] = "chisel"
```
[the remark this answers is not in the document]
Comment (Nostrademous) [reply src/Classes/TradeQuery.lua:31]: It would be more involved than you think. I started that way, then arrived at this as it was a much smaller table that's needed and a lot easier to actually maintain.

File: src/Classes/TradeQuery.lua (hunk 2 above)
Comment (Nostrademous) [reply src/Classes/TradeQuery.lua:862]: Yeah... I'm torn on that one. Due to the subscripts running in their own thread and my unfamiliarity with how they interface with shared memory made me re-use what I knew worked (aka `controls`). You need a structure that can be access from another thread and asynchronously updated. 

Welcome other approaches

File: src/Classes/TradeQuery.lua (hunk 4 above)
Comment (Nostrademous) [reply src/Classes/TradeQuery.lua:364]: FIXED

File: src/Classes/TradeQuery.lua (hunk 5 above)
Comment (Nostrademous) [reply src/Classes/TradeQuery.lua:146]: It's what Cinnabarit told me to use.

File: src/Data/Global.lua (hunk 6 above)
Comment (Nostrademous) [reply src/Data/Global.lua:289]: If you do that, it will get saved to the Build.xml by default. You would be sharing people's POESESSID as part of their build with pastebin, etc.

File: src/Classes/TradeQueryGenerator.lua [hunk 20]
```diff
@@ -0,0 +1,698 @@
...
+
+function TradeQueryGeneratorClass:InitMods()
+    local queryModFilePath = "Data/QueryMods.lua"
+
+    local file = io.open(queryModFilePath,"r")
+    if file then
+        file:close()
+        self.modData = LoadModule(queryModFilePath)
+        return
+    end
```
[the remark this answers is not in the document]
Comment (Nostrademous) [reply src/Classes/TradeQueryGenerator.lua:232]: All you have to do is delete the file... and it will auto-regenerate it. Only reason we save it is it takes a good 10-20 seconds on a decent computer and people complain when things "are slow".

File: src/Classes/TradeQueryGenerator.lua [hunk 21]
```diff
@@ -0,0 +1,698 @@
...
+
+function TradeQueryGeneratorClass:ExecuteQuery()
+    self:GenerateModWeights(self.modData["Explicit"])
+    self:GenerateModWeights(self.modData["Implicit"])
+    if self.calcContext.options.includeCorrupted then
+        self:GenerateModWeights(self.modData["Corrupted"])
+    end
+    --if self.calcContext.options.includeScourge then
+    --    self:GenerateModWeights(self.modData["Scourge"])
+    --end
```
[the remark this answers is not in the document]
Comment (Nostrademous) [reply src/Classes/TradeQueryGenerator.lua:527]: That was my original intent... but I had limited time and I broke a bunch of things when I attempted it... so I punted.

File: src/Classes/TradeQuery.lua [hunk 22]
```diff
@@ -0,0 +1,897 @@
...
+currencyConversionTradeMap["Blessing of Chayula"] = "blessing-chayula"
+currencyConversionTradeMap["Veiled Chaos Orb"] = "veiled-chaos-orb"
+currencyConversionTradeMap["Enkindling Orb"] = "enkindling-orb"
+currencyConversionTradeMap["Instilling Orb"] = "instilling-orb"
+currencyConversionTradeMap["Sacred Orb"] = "sacred-orb"
+
+
+local TradeQueryClass = newClass("TradeQuery", function(self, itemsTab)
+    self.itemsTab = itemsTab
+	self.itemsTab.leagueDropList = { }
```
Comment (Nightblade) [inline src/Classes/TradeQuery.lua:84]: I think there's some mixed tab/space indenting here?

Comment (Nostrademous) [reply src/Classes/TradeQuery.lua]: Yep... fixed

File: src/Data/Global.lua (hunk 6 above)
Comment (Wires77) [reply src/Data/Global.lua:289]: Not Build.xml, Settings.xml.  A user's settings doesn't get stored with a build.

File: src/Classes/TradeQueryGenerator.lua (hunk 21 above)
[the remark this answers is not in the document]
Comment (Wires77) [reply src/Classes/TradeQueryGenerator.lua:533]: Fair enough, not super critical to include on first pass anyway

File: src/Classes/TradeQueryGenerator.lua (hunk 20 above)
[the remark this answers is not in the document]
Comment (Wires77) [reply src/Classes/TradeQueryGenerator.lua:232]: Yeah for sure, I like saving the file off for speed.  I'm guessing these mods will get added to each time one appears on the trade site, so we can't just push it out with an update.  Maybe it's possible to store a hash of the json, and if the hash is different than the file, regenerate it (or simply add to it?  Not sure which would be better).  The only downside is requesting stats is one less request we have for actual queries

File: src/Classes/TradeQuery.lua (hunk 18 above)
[the remark this answers is not in the document]
Comment (Wires77) [reply src/Classes/TradeQuery.lua:836]: Yeah, that makes sense.  I guess I didn't realize the autoEquip doesn't replace an existing item.  I'm not sure if the two undo states are desirable (one that imports the item, one that equips it).  These lines would turn that into just one undo state:
```lua
self.itemsTab:AddItem(item, true)
self.itemsTab:SetDisplayItem()
```

File: src/Classes/TradeQuery.lua (hunk 2 above)
Comment (Wires77) [reply src/Classes/TradeQuery.lua:862]: Okay, I'll take a look once I dive into the code and see if there's another way.  Coroutines definitely make it more complex than when I first commented on it, but UI elements as data storage is pretty icky even still.

File: src/Classes/TradeQueryGenerator.lua (hunk 20 above)
[the remark this answers is not in the document]
Comment (Nostrademous) [reply src/Classes/TradeQueryGenerator.lua:232]: We should only do this approach on the PoB Development side (not as part of client distributable). To do a hash of a file you have to pull it all down. It's a hefty file.

File: src/Classes/TradeQueryGenerator.lua (hunk 21 above)
[the remark this answers is not in the document]
Comment (Nostrademous) [reply src/Classes/TradeQueryGenerator.lua]: Added it back and working for `Standard` and `Hardcore` leagues.

File: src/Classes/TradeQuery.lua [hunk 23]
```diff
@@ -212,61 +126,50 @@ end
 
 -- Method to pull down and interpret the PoE.Ninja JSON endpoint data
 function TradeQueryClass:PullPoENinjaCurrencyConversion(league, controls)
+	local now = os.time()
```
Comment (Wires77) [inline src/Classes/TradeQuery.lua:129]: This should probably be using `get_time()` since we're using it elsewhere anyway

File: src/Classes/TradeQuery.lua [hunk 24]
```diff
@@ -814,7 +494,8 @@ function TradeQueryClass:PriceItemRowDisplay(controls, str_cnt, slotTbl, top_pan
 		elseif controls['uri'..str_cnt].buf == "" then
 			activeSlotRef.pbURL = ""
 		end
-		return validURL and self:PriceBuilderCanSearch(controls) and self:PriceBuilderCanFetch(controls)
+		local isSearching = controls['priceButton'..str_cnt].label == "Searching..."
```
Comment (Wires77) [inline src/Classes/TradeQuery.lua:499]: This feels kind of hokey to me.  Might see if there's a better way to do this logic when I dig in some more.

File: src/Classes/TradeQueryRequests.lua [hunk 25]
```diff
@@ -0,0 +1,275 @@
...
+						errMsg = "Complex Query - Please provide your POESESSID"
+					elseif response.error.message then
+						errMsg = response.error.message
+					end
+				else
+					errMsg = "No Matching Results Found"
+				end
+				return callback(nil, nil, errMsg)
+			end
+			callback(response.result, response.id, errMsg)
```
Comment (Wires77) [inline src/Classes/TradeQueryRequests.lua:146]: This feels weird that we're passing individual bits of the response back to the callback.  Why not just pass the whole response and let the implementing function take care of the pieces?

File: src/Classes/TradeQueryRequests.lua [hunk 26]
```diff
@@ -0,0 +1,275 @@
...
+			end
+			local response, response_err = dkjson.decode(response)
+			if not response or not response.result then
+				local errMsg
+				if response_err then
+					errMsg = "JSON Parse Error "
+				else
+					errMsg = "Failed to Get Trade Items"
+				end
+				return callback(nil, errMsg)
```
Comment (Wires77) [inline src/Classes/TradeQueryRequests.lua:196]: This is clobbering the existing errorMsg if there was one.  Consider setting `errMsg = "JSON Parse Error: " + errMsg` without redeclaring `errMsg`

File: src/Classes/TradeQueryRequests.lua [hunk 27]
```diff
@@ -0,0 +1,275 @@
...
+			return callback(items)
+		end
+	})
+end
+
+---@param callback fun(items:table, errMsg:string)
+function TradeQueryRequestsClass:SearchWithURL(url, callback)
+	local league, queryId = url:match("https://www.pathofexile.com/trade/search/(.+)/(.+)$")
+	-- self:FetchSearchQuery(queryId,
+	-- testing experimental HTML parsing
```
Comment (Wires77) [inline src/Classes/TradeQueryRequests.lua:216]: Is this still experimental?

File: src/Classes/TradeQueryRequests.lua [hunk 28]
```diff
@@ -0,0 +1,275 @@
...
+end
+
+---EXPERIMENTAL HTML parsing to circumvent extra API call for query fetching
+--- queryId -> query fetching via Poe API call costs precious search requests
+--- But the search page HTML also contains the query object and this request is not throttled
+---@param queryId string
+---@param callback fun(query:string, errMsg:string)
+---@see TradeQueryRequests#FetchSearchQuery
+function TradeQueryRequestsClass:FetchSearchQueryHTML(queryId, callback)
+	-- the league doesn't affect query so we set it to Standard as it doesn't change
```
Comment (Wires77) [inline src/Classes/TradeQueryRequests.lua:252]: That is very interesting...

File: src/Classes/TradeQueryRequests.lua [hunk 29]
```diff
@@ -0,0 +1,275 @@
...
+	self:DownloadPage("https://www.pathofexile.com/trade/search/Standard/" .. queryId, 
+		function(response, ErrMsg)
+			if ErrMsg then
+				return callback(nil, ErrMsg)
```
Comment (Wires77) [inline src/Classes/TradeQueryRequests.lua:256]: Nitpick: I think we'd standardize on `errMsg`

**Review by Wires77 — COMMENTED** [review]
Looks alright, but is there a reason we still didn't use the Launch:DownloadPage function?  I ask because these request don't respect the global IPv4 vs. IPv6 setting, for example, among other things, like ease of maintenance

File: src/Classes/TradeQueryRequests.lua (hunk 25 above)
Comment (Nostrademous) [reply src/Classes/TradeQueryRequests.lua:146]: /shrug

Comment (Dullson) [reply src/Classes/TradeQueryRequests.lua:146]: Returning the full response just hands over the parsing responsibility to the caller. Initially, there were calls from UI code so I didn't want to leak that logic into UI (not the case anymore). Also seemed like a good idea to constrain functions as much as possible so the intent is clear. In this case, item hashes and the query id are the only required data to perform an item fetch.
Anyways, the change commit is shorter than this message if you prefer it that way. 😀

File: src/Classes/TradeQuery.lua (hunk 24 above)
Comment (Dullson) [reply src/Classes/TradeQuery.lua:499]: I didn't touch any UI stuff except a few lines and this was one of them. I had to throw in an ugly fix while testing the rate limiter. This is one of the reasons why I think a UI review is necessary before merging.

**Review by Wires77 — APPROVED** [review]
This has been out there long enough without any bugs (just a few feature requests), so I'm good with merging it into dev now.
